spec/compiler/crystal/tools/doc/type_spec.cr

```diff
@@ -44,4 +44,58 @@ describe Doc::Type do
     foo.lookup_class_method("new").should_not be_nil
     foo.lookup_class_method("new", 1).should_not be_nil
   end
+
+  describe "#node_to_html" do
+    it "shows relative path" do
+      program = semantic(<<-CODE, wants_doc: true).program
```

Comment:
Can you try not to parse a program for every sample? It's relatively costly and really not the target of this spec.
Instead, you can just instantiate the AST nodes directly:
```cr
program = Program.new
generator = Doc::Generator.new program, [""]

foo_class = NonGenericClassType.new program, program, "Foo", nil
bar_class = NonGenericClassType.new program, foo_class, "Bar", nil
bar_class.add_location Location.new("", 0, 0)
foo_class.types["Bar"] = bar_class

foo = Doc::Type.new generator, foo_class
foo.node_to_html("Bar".path).should eq(%(<a href="Foo/Bar.html">Bar</a>))
```

Reply:
Okay. I'll try it.

Reply:
I know is more unit testing if we create the AST, but they are simpler to write if we leave the parsing and also we can capture some of the issues of ordering annotation and docs that we discovered recently.

It's true that compiling the parser and the compiler is heavier that the AST + doc tool probably. I don't do that much, but I see some benefits in parsing crystal code in theses specs.

Reply:
The problem is not so much compiling the spec code because some specs need the compiler behaviour anyway, so that doesn't add any overhead. But running each spec takes a compiler pass (not full semantics, but still). Compiler specs already take more than 10 minutes to run. And I figure we're still missing a couple hundred specs to properly test the doc generator.

We obviously also need specs based on parsing source code directly as integration specs. But that's not necessary and ineffecitient for testing individual features like `#node_to_html`.

Creating spec data using AST nodes directly isn't that difficult if the API was more accessible. The definitions for AST nodes are separated over multiple files and largely undocumented. If there was a way to properly inspect a parsed AST it would also help a lot for creating it in code (ASTNode#inspect just returns the source code).

Reply:
@straight-shoota Rewriting specs are done. But I'm not sure it is better.

The following list is my feeling on rewriting specs:

  1. Building `Program` object by the hand is hard in fact. `Doc::Generator` and related classes properly work with **complete** `Program` object, even though their individual methods. However, current spec ensure only that `Doc::Type#node_to_html` works with **incomplete** `Program` object. What's this worth?
  2. Less readability. It makes hard to understand each specs.

Reply:
Yeah, true it's not really tesing isolated code and takes the entire program context into account. But that's just the nature of the method being tested and it has a benefit to explicitly state these dependencies in the spec code.

I'm not sure about the less readability argument. Yes it's hard to understand what the equivalent in stringified source code looks like (which we're more accustomed to). But manually created AST trees focus on the features that are really necesssary for the spec and make the explicit while removing unsubstantial attachments.

When working with the code that implements such specs, you need to understand the structure of the AST tree anyways. So it' may actually be helpful to understand how the method works if you don't have to follow what AST tree the parser makes out of the source code.

Maybe this specific case could be improved by separating the behaviour related to extracting information from program context and the actual result rendering. Then the specs can focus on smaller and more isolated features.

Reply:
I always prefer black box specs. Yes, the test will depend on some properties like types and doc, but if later on we want to redactor how types are named or represented then it's likely this test will still work.

Please revert to the old way you tested it which was fine. Test speed shouldn't be a reason to make a test less readable or less black boxy.

Reply:
It's reverted.